experiments/optimizer/run.py

```diff
@@ -0,0 +1,413 @@
+        self.tokenizer = GPT2TokenizerFast.from_pretrained("flax-community/gpt2-base-thai" , bos_token='<|startoftext|>', eos_token='<|endoftext|>', pad_token='<|pad|>')
+        self.max_tokens = max_tokens
+
+    def __iter__(self):
+        buffer = []
+        for sample in load_dataset("mc4",
+                       languages=["th"], 
+                        streaming=True, # optional
+                        split="train").shuffle(buffer_size=10_000):
+            buffer += self.tokenizer( sample["text"] + '<|endoftext|>')["input_ids"]
+            #buffer += [self.tokenizer.eos_token_id]
+            #print(self.tokenizer.decode(buffer))
+            while len(buffer) > self.max_tokens:
+                yield torch.tensor(buffer[: self.max_tokens])
+                buffer = buffer[self.max_tokens :]
+
+class DatasetWrapper_val(IterableDataset):
+    def __init__(self, max_tokens=256):
+        self.tokenizer = GPT2TokenizerFast.from_pretrained("flax-community/gpt2-base-thai" , bos_token='<|startoftext|>', eos_token='<|endoftext|>', pad_token='<|pad|>')
+        self.max_tokens = max_tokens
+
+    def __iter__(self):
+        buffer = []
+        for sample in load_dataset("mc4",
+                       languages=["th"], 
+                        streaming=True, # optional
+                        split="validation"):
+            buffer += self.tokenizer( sample["text"] + '<|endoftext|>')["input_ids"]
+            #buffer += [self.tokenizer.eos_token_id]
+            #print(self.tokenizer.decode(buffer))
+            while len(buffer) > self.max_tokens:
+                yield torch.tensor(buffer[: self.max_tokens])
+                buffer = buffer[self.max_tokens :]
+
+
+class Trainer:
+    def __init__(self, optimizer ,seed  , batch_size  , context_length , max_steps ,
+                      eval_steps ,  warmup_steps , model_name  , weight_decay  , grad , lr ,do_sample , use_flash ):
+        self.max_tokens = context_length
+        self.grad = grad
+        self.step = 0
+        self.max_steps = max_steps
+        self.seed = seed
+        self.warmup_steps = warmup_steps
+        self.eval_steps = eval_steps
+        self.do_sample = do_sample
+        self.dataset = DatasetWrapper(self.max_tokens)
+        self.dataset_val = DatasetWrapper_val(self.max_tokens)
+        self.use_flash = use_flash
+
+        self.tokenizer = self.dataset.tokenizer
+        self.loader = DataLoader(
+            self.dataset,
+            batch_size=batch_size,
+            num_workers=2,
+        )
+
+        self.loader_val = DataLoader(
+            self.dataset_val,
+            batch_size=batch_size
+        )
+
+        self.scaler = torch.cuda.amp.GradScaler()
+        self.model = model = make_model(model_name, self.max_tokens , self.tokenizer , self.use_flash)
+       
+        if optimizer =="lion":
+            print("Use lion optimizer")
+            self.opt = Lion(model.parameters(), lr=lr, weight_decay=weight_decay,)
+        else:
```

Comment:
I think this line should be `elif optimizer == "AdamW"` and in the `else` condition you should raise `NotImplementedError`.